d2/src/test/java/com/linkedin/d2/balancer/util/canary/TestBasicCanaryDistributionProviderImpl.java

```diff
@@ -0,0 +1,210 @@
+            .setScope(0.4);
+    D2CanaryDistributionStrategy targetAppsStrategy =
+        new D2CanaryDistributionStrategy().setStrategy(StrategyType.TARGET_APPLICATIONS)
+            .setTargetApplicationsStrategyProperties(appsProperties);
+
+    CanaryDistributionProviderImplFixture fixture = new CanaryDistributionProviderImplFixture();
+
+    // NOT in target list
+    Assert.assertEquals(fixture.getSpiedImpl("appC", null).distribute(targetAppsStrategy),
+        CanaryDistributionProvider.Distribution.STABLE, "App not in target list should return stable.");
+
+    // in scope and target list
+    Assert.assertEquals(fixture.getSpiedImpl("appA", null, 38).distribute(targetAppsStrategy),
+        CanaryDistributionProvider.Distribution.CANARY,
+        "App in target list and hash result in canary scope should return canary.");
+
+    // not in scope and in target list
+    Assert.assertEquals(fixture.getSpiedImpl("appA", null, 50).distribute(targetAppsStrategy),
+        CanaryDistributionProvider.Distribution.STABLE,
+        "App in target list but hash result not in canary scope should return stable.");
+  }
+
+  /**
+   * Provide objects for testing edge cases
+   * @return a list of objects with structure: {
+   *  D2 canary distribution strategy
+   * }
+   */
+  @DataProvider(name = "getEdgeCaseStrategies")
+  public Object[][] getEdgeCaseStrategies()
+  {
+    return new Object[][]{
+        // percentage strategy missing properties
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.PERCENTAGE)},
+        // target hosts strategy missing properties
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.TARGET_HOSTS)},
+        // target apps strategy missing properties
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.TARGET_APPLICATIONS)},
+        // strategy with disabled type
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.DISABLED)},
+        // strategy with unknown type
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.$UNKNOWN)}};
+  }
+
+  @Test(dataProvider = "getEdgeCaseStrategies")
+  public void testEdgeCases(D2CanaryDistributionStrategy strategy)
+  {
+    Assert.assertEquals(new CanaryDistributionProviderImplFixture().getSpiedImpl().distribute(strategy),
+        CanaryDistributionProvider.Distribution.STABLE, "Invalid strategies should return stable");
+  }
+
+  @Test
+  public void testHashDistribution()
+  {
+    // verify hash results are almost uniformly distributed
+    int bucketCount = 100;
+    int attemptCount = 1000;
+    int[] hits = new int[bucketCount]; // to count hits in each bucket 0 ~ 99.
+    String serviceName = "test-service";
+    String hostName = "lor1-app";
+    int bucketIdx;
+    int hashResult;
+    for (int i = 0; i < attemptCount; i++) {
+      hashResult = (serviceName + hostName + i).hashCode();
```

Comment:
This should be using the actual implementation in BasicCanaryDistributionProviderImpl, otherwise that might change and this test will still pass.

Reply:
makes sense. fixed.